src/python/grpcio_tests/tests/unit/BUILD.bazel

```diff
@@ -84,29 +84,39 @@ py_library(
 )
 
 [
-    py2and3_test(
-        name=test_file_name[:-3],
-        size="small",
-        srcs=[test_file_name],
-        main=test_file_name,
-        deps=[
-            "//src/python/grpcio/grpc:grpcio",
-            ":resources",
-            ":test_common",
-            ":thread_pool",
-            ":_exit_scenarios",
-            ":_server_shutdown_scenarios",
-            ":_from_grpc_import_star",
-            ":_tcp_proxy",
-            ":_signal_client",
-            "//src/python/grpcio_tests/tests/unit/framework/common",
-            "//src/python/grpcio_tests/tests/testing",
-            "//external:six"
-        ],
-        imports=["../../",],
-        data=[
-            "//src/python/grpcio_tests/tests/unit/credentials",
-        ],
-    ) for test_file_name in GRPCIO_TESTS_UNIT
+    [
+        py2and3_test(
+            name=test_file_name[:-3],
+            size="small",
+            srcs=[test_file_name],
+            main=test_file_name,
+            deps=[
+                "//src/python/grpcio/grpc:grpcio",
+                ":resources",
+                ":test_common",
+                ":thread_pool",
+                ":_exit_scenarios",
+                ":_server_shutdown_scenarios",
+                ":_from_grpc_import_star",
+                ":_tcp_proxy",
+                ":_signal_client",
+                "//src/python/grpcio_tests/tests/unit/framework/common",
+                "//src/python/grpcio_tests/tests/testing",
+                "//external:six"
+            ],
+            imports=["../../",],
+            data=[
+                "//src/python/grpcio_tests/tests/unit/credentials",
+            ],
+        ),
+        sh_test(
+            name = test_file_name[:-3] + ".single_threaded_unary_stream",
+            data = ["{}.python3".format(test_file_name[:-3])],
+            args = ["{}.python3".format(test_file_name[:-3])],
+            size = "small",
+            srcs = ["_single_threaded_unary_stream_test_runner.sh"]
+        )
```

Comment:
I might not considered all cases here, but I have two thoughts for this approach:

1. Why don't we use the `--test_env` for the Bazel testing script? It runs the tests for us automatically, with minimum amount of code changes?
2. If we do want to use `sh_test`, can we split them into multiple for loops in this BUILD file? I hope they can be decoupled with existing `py2and3_test`s.

Reply:
1. I don't like this option because it means that `bazel test //src/python/...` will not pick up these tests.
2. Sure. The difference is purely syntactic though. There's an inherent coupling here. The idea is to run every existing test with the new code path.

Reply:
Personally, I prefer the `--test_env` because of simplicity.

1. Using existed mechanism is more readable and introduces less maintenance burden to future ourselves.
2. Coverage is still guaranteed. We can add the `--test_env` to our CI script, so Kokoro would run it for us.
3. For development speed, we can add one more Bazel config mode in `.bazelrc` to specify this behavior, so we can type less when we run this suite of tests.
4. This might work against Bazel's design philosophy? That environment variables should apply to all libraries so all of them behave consistently.

---

Alternatively, I have another idea that since this is a workaround for tests only, I think we can use certain mechanism to accept command line arguments instead of environment variable.

E.g. we can add a Python testing module that checks command line argument in the import phase, and flip the flag in gRPC module. So, it satisfies both your needs and my needs.